src/common/heartbeatService.js

```diff
@@ -0,0 +1,104 @@
+const {
+  APPLICATION_TAG_KEY,
+  CLUSTER_TAG_KEY,
+  SERVICE_TAG_KEY,
+  SHARD_TAG_KEY,
+  COMPONENT_TAG_KEY,
+  HEART_BEAT_METRIC,
+  HEART_BEAT_INTERVAL,
+  NULL_TAG_VAL
+} = require('./constants');
+
+/**
+ * Service that periodically reports component heartbeats to Wavefront.
+ */
+class HeartbeatService {
+  /**
+   *
+   * @param {WavefrontDirectClient} wavefrontClient
+   * @param applicationTags
+   * @param {Array} components - Array of string indicates Components.
+   * @param source
+   */
+  constructor(wavefrontClient, applicationTags, components, source) {
+    this.wavefrontClient = wavefrontClient;
+    this.applicationTags = applicationTags;
+    this.source = source;
+    this.reportingIntervalSeconds = HEART_BEAT_INTERVAL;
+    this.heartbeatMetricTagsList = [];
+    this.customTagsSet = new Set();
+
+    if (!Array.isArray(components)) {
+      components = [components];
+    }
+
+    for (let component of components) {
+      let metricTags = {
+        APPLICATION_TAG_KEY: applicationTags.application,
+        CLUSTER_TAG_KEY: applicationTags.cluster || NULL_TAG_VAL,
+        SERVICE_TAG_KEY: applicationTags.service,
+        SHARD_TAG_KEY: applicationTags.shard || NULL_TAG_VAL,
+        COMPONENT_TAG_KEY: component
+      };
+
+      if (applicationTags.customTags) {
+        metricTags = { ...metricTags, ...applicationTags.customTags };
+      }
+      this.heartbeatMetricTagsList.push(metricTags);
+    }
+    this._timer = null;
+    this._run();
+  }
+
+  /**
+   * Schedule the timer to report.
+   */
+  _run() {
+    this._timer = setInterval(this._report, this.reportingIntervalSeconds);
+  }
+
+  _report() {
+    try {
+      this.heartbeatMetricTagsList.forEach(heartbeat => {
+        this.wavefrontClient.sendMetric(
+          HEART_BEAT_METRIC,
+          1.0,
+          Date.now(),
+          this.source,
+          heartbeat
+        );
+      });
+      this.customTagsSet.forEach(customTagsMap => {
+        this.wavefrontClient.sendMetric(
+          HEART_BEAT_METRIC,
+          1.0,
+          Date.now(),
+          this.source,
+          customTagsMap
+        );
+        this.customTagsSet.delete(customTagsMap);
```

Comment:
I'm not sure if this operation is safe. You are deleting an element from the same set you're iterating over.

Reply:
that's a good point. I found JS Set safe about deletion while iterating because essentially it does `Replacing it with an empty value rather than deleting it ensures it will not mess up with iterators' positions`.
(https://stackoverflow.com/questions/28306756/is-it-safe-to-delete-elements-in-a-set-while-iterating-with-for-of)